### PR TITLE
Helm: Improve experience for editing chart values

### PR DESCRIPTION
**Pull request by nwmac**
Currently we use a very basic text area for editing values when installing/upgrading Helm charts/releases.

This PR improves the UI for this:

- Uses the Monaco editor for better code editing
- Uses monaco-yaml for yaml language support
- Uses the JSON schema form library to present a form when a JSON Schema is available
- Loads JSON Schemes/values into monaco-yaml to provide code completion in the editor

The editor component also now sizes properly to correctly used the available space.

### CODE REVIEW

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/chart-values-editor/chart-values-editor.component.ts [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+        this.initialFormData = data;
+        this.formData = data;
+      } catch (e) {
+        // The yaml in the code editor is invalid, so we can't marshal it back to json for the from editor
+        this.yamlError = true;
+      }
+    }
+  }
+
+  // Called once the Monaco editor has loaded - store a reference to the editor and ensure the editor theme
```
Comment (nwmac) [inline src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/chart-values-editor/chart-values-editor.component.ts]: Called multiple times - each time model changes

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/chart-values-editor/chart-values-editor.component.ts [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+    // Only load the YAML support once - when we set the model, onMonacoInit will et
+    if (this.model) {
+      return;
+    }
+
+    // Load the YAML Language support - require is available as it will have been loaded by the Monaco vs loader
+    const req = (window as any).require;
+    req(['vs/language/yaml/monaco.contribution'], () => {
+      // Set the model now that YAML support is loaded - this will update the editor correctly
+      this.model = {
```
Comment (nwmac) [inline src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/chart-values-editor/chart-values-editor.component.ts]: could use updateModel method

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/chart-values-editor/chart-values-editor.component.ts [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+    const w = this.elRef.nativeElement.offsetWidth;
+    let h = this.elRef.nativeElement.offsetHeight;
+
+    // Check if host element not visible (does not have a size)
+    if ((w === 0) && (h === 0)) {
+      return;
+    }
+
+    // Remove height of toolbar (since this is incluced in the height of the host element)
+    h = h - TOOLABR_HEIGHT;
```
Comment (nwmac) [inline src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/chart-values-editor/chart-values-editor.component.ts]: typo

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/chart-values-editor/chart-values-editor.component.ts [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+    this.renderer.setStyle(this.monacoEditor.nativeElement, 'width', `${w}px`);
+    this.renderer.setStyle(this.monacoEditor.nativeElement, 'height', `${h}px`);
+
+    // Ask Monaco to layout again with its new size
+    this.editor.layout();
+  }
+
+  // Get an absolute URI for the Schema - it is not fetched, just used as a reference
+  getSchemaUri(): string {
+    return `https://stratos.app/schemas${this.schemaUrl}`;
```
Comment (nwmac) [inline src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/chart-values-editor/chart-values-editor.component.ts:337]: Can just use schema URL

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/chart-values-editor/diffvalues.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+      if(src[key] === null && dest[key] === null) {
+        delete src[key];
+      } else if(Array.isArray(src[key])) {
+        // Array
+        if (arraysAreEqual(src[key], dest[key])) {
+          delete src[key];
+        }
+      } else if (typeof(src[key]) === 'object') {
+        // Object
+        const updated = diffObjects(src[key], dest[key]);
```
Comment (nwmac) [inline src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/chart-values-editor/diffvalues.ts]: Don't need to store updated

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/chart-values-editor/diffvalues.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        return false;
+      }
+    } else if (a1[i] !== a2[i]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// diffObjects is main export - diffs two objects and returns only the diffrence
```
Comment (nwmac) [inline src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/chart-values-editor/diffvalues.ts:23]: Comment that we are only diffing one way

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/upgrade-release/upgrade-release.component.ts [hunk 7]
```diff
@@ -34,29 +38,30 @@ import { ReleaseUpgradeVersionsListConfig } from './release-version-list-config'
...
+
+  public config: ChartValuesConfig;
 
   private monocularEndpointId: string;
 
   // Future
   public showAdvancedOptions = false;
 
+  // Release metadata
+  private release: HelmRelease;
```
Comment (nwmac) [inline src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/upgrade-release/upgrade-release.component.ts]: Not sure this is needed?

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/chart-values-editor/diffvalues.ts (hunk 6 above)
Comment (nwmac) [reply src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/chart-values-editor/diffvalues.ts:23]: Done

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/chart-values-editor/diffvalues.ts (hunk 5 above)
Comment (nwmac) [reply src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/chart-values-editor/diffvalues.ts]: Removed

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/chart-values-editor/chart-values-editor.component.ts (hunk 4 above)
Comment (nwmac) [reply src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/chart-values-editor/chart-values-editor.component.ts:337]: schemaUrl is relative - so this is needed- I've added a comment

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/chart-values-editor/chart-values-editor.component.ts (hunk 3 above)
Comment (nwmac) [reply src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/chart-values-editor/chart-values-editor.component.ts]: fixed

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/chart-values-editor/chart-values-editor.component.ts (hunk 2 above)
Comment (nwmac) [reply src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/chart-values-editor/chart-values-editor.component.ts]: done

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/chart-values-editor/chart-values-editor.component.ts (hunk 1 above)
Comment (nwmac) [reply src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/chart-values-editor/chart-values-editor.component.ts]: added comment

**Review by richard-cox — APPROVED** [review]

**Pull request merged**